<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Examples &#8212; flintyPy 0.1.17 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/bootstrap-sphinx.css" />
    <link rel="stylesheet" type="text/css" href="_static/css/css-style.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">
<script type="text/javascript" src="_static/js/jquery-1.11.0.min.js "></script>
<script type="text/javascript" src="_static/js/jquery-fix.js "></script>
<script type="text/javascript" src="_static/bootstrap-3.3.7/js/bootstrap.min.js "></script>
<script type="text/javascript" src="_static/bootstrap-sphinx.js "></script>

  </head><body>

  <div id="navbar" class="navbar navbar-inverse navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="index.html">
          flintyPy</a>
        <span class="navbar-text navbar-version pull-left"><b></b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="#">Examples</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="index.html">Articles <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"></ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Contents <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">Examples</a><ul>
<li><a class="reference internal" href="#a-simulated-binary-matrix">A Simulated Binary Matrix</a></li>
<li><a class="reference internal" href="#a-real-valued-matrix">A Real-Valued Matrix</a></li>
<li><a class="reference internal" href="#a-list-of-independent-distances">A List of Independent Distances</a></li>
<li><a class="reference internal" href="#genomes-project-data">1000 Genomes Project Data</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
              
                
              
            
            
            
            
              <li class="hidden-sm">
<div id="sourcelink">
  <a href="_sources/examples.rst.txt"
     rel="nofollow">Source</a>
</div></li>
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="body col-md-12 content" role="main">
      
  <div class="section" id="examples">
<h1>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h1>
<p>Here are some examples of how <strong>flintyPy</strong> can be run on data. We are working on expanding our Python documentation, so please take a look at <a class="reference external" href="https://alanaw1.github.io/flintyR/articles">these articles</a> from our companion flintyR package if you are looking for more real-life examples. There, you will find many pedagogical examples accompanied by thorough explorations.</p>
<div class="section" id="a-simulated-binary-matrix">
<h2>A Simulated Binary Matrix<a class="headerlink" href="#a-simulated-binary-matrix" title="Permalink to this headline">¶</a></h2>
<p>Here, we present the simplest example of running our test on a binary matrix with independent features.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import modules</span>
<span class="kn">import</span> <span class="nn">flintypy</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="c1"># Simulate binary matrix</span>
<span class="n">X</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">50</span><span class="p">,</span> <span class="mi">100</span><span class="p">))</span>

<span class="c1"># Run test</span>
<span class="n">flintypy</span><span class="o">.</span><span class="n">v_stat</span><span class="o">.</span><span class="n">get_p_value</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="a-real-valued-matrix">
<h2>A Real-Valued Matrix<a class="headerlink" href="#a-real-valued-matrix" title="Permalink to this headline">¶</a></h2>
<p>A slightly more complicated example is when features are partitioned into independent blocks. Here, we simulate one such array. Features are drawn from an autoregressive (<a class="reference external" href="https://online.stat.psu.edu/stat501/lesson/14/14.1">AR</a>) process one block at a time, with successive blocks concatenated one after another. We use <span class="math notranslate nohighlight">\(B=20\)</span> blocks, with each block containing <span class="math notranslate nohighlight">\(100\)</span> draws from the AR(1) process. We sample <span class="math notranslate nohighlight">\(N=10\)</span> “individuals”.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import modules</span>
<span class="kn">import</span> <span class="nn">flintypy</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">statsmodels.tsa.arima_process</span> <span class="kn">import</span> <span class="n">ArmaProcess</span>

<span class="c1"># Define ARMA process object</span>
<span class="n">ar</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.9</span><span class="p">])</span>
<span class="n">ma</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">])</span>
<span class="n">AR_object</span> <span class="o">=</span> <span class="n">ArmaProcess</span><span class="p">(</span><span class="n">ar</span><span class="p">,</span> <span class="n">ma</span><span class="p">)</span>

<span class="c1"># Simulate blocks of AR process features</span>
<span class="n">N_BLOCKS</span> <span class="o">=</span> <span class="mi">20</span>
<span class="n">big_array</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">AR_object</span><span class="o">.</span><span class="n">generate_sample</span><span class="p">(</span><span class="n">nsample</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">)])</span> <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">N_BLOCKS</span><span class="p">)])</span>
<span class="n">X</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">(</span><span class="n">big_array</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">blocks</span> <span class="o">=</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">b</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span> <span class="n">b</span> <span class="o">*</span> <span class="mi">100</span> <span class="o">+</span> <span class="mi">100</span><span class="p">)</span> <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">N_BLOCKS</span><span class="p">)]</span>

<span class="c1"># Run test without accounting for block dependency</span>
<span class="c1"># Use large P asymptotics since P = 20 * 100 is large</span>
<span class="n">flintypy</span><span class="o">.</span><span class="n">v_stat</span><span class="o">.</span><span class="n">get_p_value</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">large_p</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>

<span class="c1"># Run test accounting for block dependency</span>
<span class="c1"># Use permutation test since B = 20 is small</span>
<span class="n">flintypy</span><span class="o">.</span><span class="n">v_stat</span><span class="o">.</span><span class="n">get_p_value</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">blocks</span> <span class="o">=</span> <span class="n">blocks</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="a-list-of-independent-distances">
<h2>A List of Independent Distances<a class="headerlink" href="#a-list-of-independent-distances" title="Permalink to this headline">¶</a></h2>
<p>Our test of exchangeability works in settings where only pairwise distance data is available. Given a list of pairwise distances between <span class="math notranslate nohighlight">\(N\)</span> individuals, <span class="math notranslate nohighlight">\(\{D_1,\ldots,D_B\}\)</span>, and assuming these distances are independent of one another, our test can determine whether the individuals making up the <span class="math notranslate nohighlight">\(N\)</span>-sample are exchangeable at a user-specified significance threshold (e.g., <span class="math notranslate nohighlight">\(\alpha = 0.05\)</span>).</p>
<p>There are multiple practical scenarios well suited for such an approach.</p>
<ul class="simple">
<li><p>Loading the individual-by-feature matrix <span class="math notranslate nohighlight">\(\mathbf{X}\)</span> into memory is computationally unfeasible</p></li>
<li><p>Only distance data is available owing to privacy issues</p></li>
<li><p>Independent, large sets of features exist on different computers but need to be combined to form the individual-by-feature matrix <span class="math notranslate nohighlight">\(\mathbf{X}\)</span></p></li>
</ul>
<p>For illustration purposes, we generate the same dataset <span class="math notranslate nohighlight">\(\mathbf{X}\)</span> using the AR(1) process described in the preceding Section. However, we assume that we only have pairwise Euclidean distances between the individuals at each block, and run the version of our test that takes in pairwise distance data, <code class="docutils literal notranslate"><span class="pre">dist_data_p_value()</span></code>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import modules</span>
<span class="kn">import</span> <span class="nn">flintypy</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">scipy.spatial.distance</span> <span class="kn">import</span> <span class="n">pdist</span>
<span class="kn">from</span> <span class="nn">scipy.spatial.distance</span> <span class="kn">import</span> <span class="n">squareform</span>
<span class="kn">from</span> <span class="nn">statsmodels.tsa.arima_process</span> <span class="kn">import</span> <span class="n">ArmaProcess</span>

<span class="c1"># Define ARMA process object</span>
<span class="n">ar</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.9</span><span class="p">])</span>
<span class="n">ma</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">])</span>
<span class="n">AR_object</span> <span class="o">=</span> <span class="n">ArmaProcess</span><span class="p">(</span><span class="n">ar</span><span class="p">,</span> <span class="n">ma</span><span class="p">)</span>

<span class="c1"># Simulate blocks of AR process features</span>
<span class="n">N_BLOCKS</span> <span class="o">=</span> <span class="mi">20</span>
<span class="n">matrix_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">AR_object</span><span class="o">.</span><span class="n">generate_sample</span><span class="p">(</span><span class="n">nsample</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">)])</span> <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">N_BLOCKS</span><span class="p">)]</span>

<span class="c1"># Convert to list of pairwise distances</span>
<span class="n">dist_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">pdist</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">metric</span> <span class="o">=</span> <span class="s1">&#39;euclidean&#39;</span><span class="p">),</span> <span class="n">matrix_list</span><span class="p">))</span> <span class="c1"># in vector form</span>
<span class="n">dist_mat_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">squareform</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="n">dist_list</span><span class="p">))</span> <span class="c1"># in square matrix form</span>

<span class="c1"># Run test on pairwise distance data</span>
<span class="n">flintypy</span><span class="o">.</span><span class="n">v_stat</span><span class="o">.</span><span class="n">dist_data_p_value</span><span class="p">(</span><span class="n">dist_list</span><span class="p">)</span>

<span class="c1"># Run test on pairwise distance matrix data</span>
<span class="n">flintypy</span><span class="o">.</span><span class="n">v_stat</span><span class="o">.</span><span class="n">dist_data_p_value</span><span class="p">(</span><span class="n">dist_mat_list</span><span class="p">)</span>
</pre></div>
</div>
<p>If you run on pairwise distance data you should see the following output:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;</span>&gt;&gt; flintypy.v_stat.dist_data_p_value<span class="o">(</span>dist_list<span class="o">)</span>
<span class="go">Distances are in vector form</span>
<span class="go">0.954</span>
</pre></div>
</div>
<p>Or if you run on pairwise distance matrix data you should see this:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;</span>&gt;&gt; flintypy.v_stat.dist_data_p_value<span class="o">(</span>dist_mat_list<span class="o">)</span>
<span class="go">Distances are in matrix form</span>
<span class="go">0.966</span>
</pre></div>
</div>
</div>
<div class="section" id="genomes-project-data">
<h2>1000 Genomes Project Data<a class="headerlink" href="#genomes-project-data" title="Permalink to this headline">¶</a></h2>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Here, we demonstrate the use of our test</p>
<ul class="simple">
<li><p>on a real dataset,</p></li>
<li><p>through embedding it within a terminal-friendly script.</p></li>
</ul>
<p>Links to the scripts are provided, and we encourage users to modify and use them for their own purposes.</p>
</div>
<p>The 1000 Genomes project (<a class="reference external" href="https://www.nature.com/articles/nature15393">The 1000 Genomes Project Consortium, 2015</a>) consists of individual genomes spanning <span class="math notranslate nohighlight">\(26\)</span> populations and <span class="math notranslate nohighlight">\(5\)</span> superpopulations. We shall test if each of the <span class="math notranslate nohighlight">\(26\)</span> populations is exchangeable.</p>
<p>There are some challenges associated with this task.</p>
<ol class="arabic simple">
<li><p>There are millions of features (called “polymorphic variants”), which makes loading the entire dataset at once memory-intensive.</p></li>
<li><p>The input files are in BIM/BED/FAM or VCF format, requiring the use of special modules that can handle such files.</p></li>
</ol>
<p>To overcome these challenges, we devise a strategy around how we want to run our test. We shall assume variants lying in different chromosomes are independent of one another. Since there are <span class="math notranslate nohighlight">\(22\)</span> automosomal chromosomes, this means that our features will be grouped into <span class="math notranslate nohighlight">\(22\)</span> large blocks. Moreover, there are packages like <strong>bed-reader</strong> (<a class="reference external" href="https://fastlmm.github.io/bed-reader/">link</a>) and <strong>pybedtools</strong> (<a class="reference external" href="https://daler.github.io/pybedtools/">link</a>) that specialise in reading genomics data files. Thus, we can accomplish our task as follows.</p>
<ol class="arabic simple">
<li><p>For each population, split the genome into <span class="math notranslate nohighlight">\(22\)</span> autosomes.</p></li>
<li><dl class="simple">
<dt>Per autosome:</dt><dd><ul class="simple">
<li><p>load it into Python using <strong>bed-reader</strong> and construct its distance matrix (<em>addresses challenge 2</em>)</p></li>
<li><p>remove the loaded autosome object from memory (<em>addresses challenge 1</em>)</p></li>
</ul>
</dd>
</dl>
</li>
<li><p>Run test of exchangeability on list of distance matrices</p></li>
</ol>
<p>We provide <a class="reference external" href="https://github.com/alanaw1/exchangeability_test/blob/main/data-analysis/1KG/test_exchangeability.py">here</a> a Python script, and <a class="reference external" href="https://github.com/alanaw1/exchangeability_test/blob/main/data-analysis/1KG/1kg_exchange_test_python.bash">here</a> a “master” bash script, for performing the steps above. Our <a class="reference external" href="https://github.com/alanaw1/exchangeability_test/blob/main/data-analysis/1KG/README.md">README</a> markdown on Github provides guidance on reproducing our results, which we report below.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 75%" />
<col style="width: 25%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><em>Population</em></p></td>
<td><p><em>p-value</em></p></td>
</tr>
<tr class="row-even"><td><p>British in England and Scotland (GBR)</p></td>
<td><p>0.000</p></td>
</tr>
<tr class="row-odd"><td><p>Finnish in Finland (FIN)</p></td>
<td><p>0.000</p></td>
</tr>
<tr class="row-even"><td><p>Southern Han Chinese, China (CHS)</p></td>
<td><p>0.000</p></td>
</tr>
<tr class="row-odd"><td><p>Puerto Rican in Puerto Rico (PUR)</p></td>
<td><p>0.000</p></td>
</tr>
<tr class="row-even"><td><p>Chinese Dai in Xishuangbanna, China (CDX)</p></td>
<td><p>0.000</p></td>
</tr>
<tr class="row-odd"><td><p>Colombian in Medellin, Colombia (CLM)</p></td>
<td><p>0.000</p></td>
</tr>
<tr class="row-even"><td><p>Iberian populations in Spain (IBS)</p></td>
<td><p>0.000</p></td>
</tr>
<tr class="row-odd"><td><p>Peruvian in Lima, Peru (PEL)</p></td>
<td><p>0.000</p></td>
</tr>
<tr class="row-even"><td><p>Punjabi in Lahore, Pakistan (PJL)</p></td>
<td><p>0.000</p></td>
</tr>
<tr class="row-odd"><td><p>Kinh in Ho Chi Minh City, Vietnam (KHV)</p></td>
<td><p>0.000</p></td>
</tr>
<tr class="row-even"><td><p>African Caribbean in Barbados (ACB)</p></td>
<td><p>0.000</p></td>
</tr>
<tr class="row-odd"><td><p>Gambian in Western Division, The Gambia (GWD)</p></td>
<td><p>0.000</p></td>
</tr>
<tr class="row-even"><td><p>Esan in Nigeria (ESN)</p></td>
<td><p>0.000</p></td>
</tr>
<tr class="row-odd"><td><p>Bengali in Bangladesh (BEB)</p></td>
<td><p>0.000</p></td>
</tr>
<tr class="row-even"><td><p>Mende in Sierra Leone (MSL)</p></td>
<td><p>0.000</p></td>
</tr>
<tr class="row-odd"><td><p>Sri Lankan Tamil in the UK (STU)</p></td>
<td><p>0.000</p></td>
</tr>
<tr class="row-even"><td><p>Indian Telugu in the UK (ITU)</p></td>
<td><p>0.000</p></td>
</tr>
<tr class="row-odd"><td><p>Utah residents (CEU)</p></td>
<td><p>0.000</p></td>
</tr>
<tr class="row-even"><td><p>Yoruba in Ibadan, Nigeria (YRI)</p></td>
<td><p>0.029</p></td>
</tr>
<tr class="row-odd"><td><p>Han Chinese in Beijing, China (CHB)</p></td>
<td><p>0.000</p></td>
</tr>
<tr class="row-even"><td><p>Japanese in Tokyo, Japan (JPT)</p></td>
<td><p>0.000</p></td>
</tr>
<tr class="row-odd"><td><p>Luhya in Webuye, Kenya (LWK)</p></td>
<td><p>0.000</p></td>
</tr>
<tr class="row-even"><td><p>African Ancestry in Southwest US (ASW)</p></td>
<td><p>0.000</p></td>
</tr>
<tr class="row-odd"><td><p>Mexican Ancestry in Los Angeles, California (MXL)</p></td>
<td><p>0.000</p></td>
</tr>
<tr class="row-even"><td><p>Toscani in Italy (TSI)</p></td>
<td><p>0.000</p></td>
</tr>
<tr class="row-odd"><td><p>Gujarati Indian in Houston, TX (GIH)</p></td>
<td><p>0.000</p></td>
</tr>
</tbody>
</table>
<p>These numbers tell us that all <span class="math notranslate nohighlight">\(26\)</span> populations have non-exchangeable samples. (For geneticists: note that we have included rare variants with minor allele count &gt;= 1 in our analysis. See <a class="reference external" href="https://alanaw1.github.io/flintyR/articles/extras.html#running-our-test-from-terminal-1">this vignette</a>, <strong>Preparing the population-level files</strong>, for details.) This observation is consistent with the importance of rare variants capturing recent demographic history, a point acknowledged by population geneticists (<a class="reference external" href="https://elifesciences.org/articles/61548">Zaidi and Mathieson, 2021</a>).</p>
</div>
</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2021, Alan Aw, Jeffrey Spence and Yun Song.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 4.1.2.<br/>
    </p>
  </div>
</footer>
  </body>
</html>